File: fod.js
'use strict';

async function fetchScans(ids) {
    let t = await fod.authenticate.bySecret()     //Authenticate (must be done synchronously before other REST calls)
    var x = {"id": ids}  // Scan Id's to capture
    var s = []  // holds the summary results
    var m = []  // holds the manifest results
    x.id.forEach(id => {
        s.push(fod.scan.getSummary(id))
        m.push(fod.scan.getManifest(id))
    });
    var a = []  // Promise container to make sure both summaries and manifests fetches have all completed
    a.push(Promise.all(s).then(v => {   // Summaries Promise wrapper - once all are fetched, compare for differences
        x["data"]=v
        x["differences"] = differences(v)
    }))
    a.push(Promise.all(m).then(v => {   // Manifests Promise wrapper - once all are fetched, parse and merge results
        for (var i=0; i<v.length; i++) {
            v[i]=parse(v[i])
        }
        x["manifest"]=merge(v)
    }))
    Promise.all(a).then(v => {  // Once all fetches are done, export to json and excel
        const fs = require('fs')
        fs.writeFileSync("differences.json", JSON.stringify(x))

        var xl = require('excel4node');     // Excel builder
        var wb = new xl.Workbook();
        var ws = wb.addWorksheet('Differences');    // Write differences to tab1
        ws.cell(1,1).string("Property")
        var j = 2
        Object.keys(x.differences).forEach(k => {   // iterate through keys
            for (var i=0; i<x.id.length; i++) {
                if (i==0) ws.cell(j,1).string(`${k}`)
                if (j==2) ws.cell(1,i+2).string(`ScanId: ${x.id[i]}`)
                var v = x.differences[k][i] // cell value
                var cl = ws.cell(j,i+2)
                if (v!=null) {
                    switch (typeof(v)) {
                        case "number": cl.number(v); break;
                        case "string": cl.string(v); break;
                        default: cl.string(typeof(v))   // this is non-string or non-number
                    }
                }
            }
            j++
        })
        var ws = wb.addWorksheet('Manifest');   // Write manifest to tab2
        ws.cell(1,1).string("Path")
        var j = 2
        Object.keys(x.manifest).forEach(k => {  // key iterator
            for (var i=0; i<x.id.length; i++) {
                if (i==0) ws.cell(j,1).string(`${k}`)
                if (j==2) {
                    ws.cell(1,i+2).string(`Sz: ${x.id[i]}`)     // We've got 2 sets of columns times # of ids (2*id.length)
                    ws.cell(1,i+x.id.length+2).string(`Dt: ${x.id[i]}`)
                }
                for (var l=0; l<2; l++) {   // column sets with l loop
                    var n = i+x.id.length*l
                    var v = x.manifest[k][n]
                    var cl = ws.cell(j,n+2)
                    if (v!=null) {
                        switch (typeof(v)) {
                            case "number": cl.number(v); break;
                            case "string": cl.string(v); break;
                            default: cl.string(typeof(v))
                        }
                    }
                }
            }
            j++
        })
        wb.write("differences.xlsx")
    })
}

function differences(summary, prefix="") {
    var r = {}
    var x = summary[0]  // work through keys of first summary object
    if (x!=null) {
        Object.keys(x).forEach(k => {   // loop over all properties
            var v = []
            var m = true    // set match flag as true
            var l = x[k]    // initial item property value
            summary.forEach(s => {  // for each key, check if all values match
                v.push(s[k])
                if (s[k]!=l) m = false  // any value not matching flags to false
            })
            if (typeof(l)!="object") {  // if the property is not an "object" type (i.e. is a primitive, check to see if all values match)
                if (!m) r[`${prefix}${k}`]=v    // add to differences if match was flagged as false
            } else {
                var z = differences(v, `${prefix}${k}.`)    // for properties that are an object (child object), recurse through child properties
                Object.keys(z).forEach(w => {
                    r[w]=z[w]   // prefix makes sure to perpetuate the property names with dots
                })
            }
        })
    }
    return r
}
function parse(manifest) {  // parses manifests in the format:  "<filepath>" size date
    var r = []
    var l = manifest.trim().split("\n") // split into lines on linefeed char
    for (var i=0; i<l.length; i++) {
        var f = l[i].match(/(".*?"|[^"\s]+)(?=\s*|\s*$)/g); // regex to ignore doublequoted text and split on whitespace as delimiters
        f[0]=f[0].slice(1,-1)   // slice off doublequotes from first item
        f[1]=Number(f[1])   // turn second field into a number
        r.push(f)
    }
    return r
}
function merge(manifests) {
    var r = {}
    var i = 0;
    for (var i=0; i<manifests.length; i++) {    // loop through manifests
        manifests[i].forEach(l => {     // loop through lines within manifest
            var p = l[0]    // first item of array is the path -- functions as keyname
            if (r[p]==null) r[p] = new Array((manifests.length*2)).fill(null)   // if key not present yet, initialize array of nulls
            r[p][i]=l[1]    // keep all field 1's (size?) together
            r[p][(manifests.length+i)]=l[2] // keep all field 2's (date) together
        })
    }
    return r
}

const fod = require("./fod-api.js");    //include CommonJS module+
fod.config = require("./.config.json")  ///load the configuration from JSON (be sure to edit for your settings)


// Fetch comparisons of these scans
fetchScans([6395564, 6395413, 6386441])

File: fod.py
import json
import requests
import csv
import xlwt

# Fetch config from config.json
def config():
    with open('.config.json') as f:
        return json.load(f)
    # return something like this:
    # {
    #     "url": "https://api.ams.fortify.com",
    #     "username": "******",
    #     "password": "******",
    #     "tenant": "tenantname"
    # }
    
# Authenticate
def authenticate(config):
    c = config
    # can use either password or client credentials -- comment out one of the next two lines accordingly
    # with requests.post(f"{c['url']}/oauth/token", data={"grant_type":"client_credentials","scope":"api-tenant","client_id":f"{c['apikey']}","client_secret":f"{c['secret']}"}) as r:
    with requests.post(f"{c['url']}/oauth/token", data={"grant_type":"password","scope":"api-tenant","username":f"{c['tenant']}\{c['username']}","password":f"{c['password']}","security_code":""}) as r:
        return json.loads(r.text)
        
# Fetch scan summary through REST
#       Example: https://api.ams.fortify.com/api/v3/scans/6395564/summary
def scansummary(config, authentication, id):
    c = config
    a = authentication
    with requests.get(f"{c['url']}/api/v3/scans/{id}/summary", headers={"authorization":f"Bearer {a['access_token']}"}) as r:
        return json.loads(r.text)

# Fetch manifest through REST
#       Example: https://api.ams.fortify.com/api/v3/scans/6395564/manifest
def manifest(config, authentication, id):
    c = config
    a = authentication
    with requests.get(f"{c['url']}/api/v3/scans/{id}/manifest", headers={"authorization":f"Bearer {a['access_token']}"}) as r:
        return r.text.strip()

# This makes a dictionary where property values differ between runs
def differences(scansummary, prefix: str=""):
    r = {}
    x = scansummary[0]
    for k in x:
        v = []
        m = True
        l = x.get(k)
        for y in scansummary:
            z = y.get(k)
            v.append(z)
            if z!=l: m = False
        if isinstance(l, str) or not hasattr(l,'__iter__'):     # check isString or not-iterable type (primitives like int, bool, etc)
            if m==False:     # only append to results if different
                r[f"{prefix}{k}"] = v
        else: 
            z = differences(v, f"{prefix}{k}.")  # for non-string iterables recursively call this function, appending key prefix
            for w in z:
                r[w]=z.get(w)
    return r

# Makes a dictionary of all manifest files -- columns of sizes & dates for comparison
def merge(manifest):
    r = {}
    i = 0
    for y in manifest:
        for l in csv.reader(y.splitlines(), delimiter = ' ', skipinitialspace=True):
            p = l[0]
            if r.get(p) is None:
                r[p]=[None] * (len(manifest) * 2)

            r[p][i]=int(l[1])               # keep all field 1's (size?) together
            r[p][(len(manifest)+i)]=l[2]    # keep all field 2's (date) together
        i+=1
    return r

def fetchScans(ids):
    # Get config details and logon
    c = config()
    a = authenticate(c)

    # Setup a variable to hold our ScanIds and the rest outputs
    x = {"id": ids}
    s = []
    m = []
    for id in x["id"]:
        s.append(scansummary(c,a,id))
        m.append(manifest(c,a,id))
    x["data"]=s
    x["differences"]=differences(s)     # Enumerate the differences
    x["manifest"]=merge(m)              # merge the manifests

    # Spool results to JSON file
    with open(f"differences.json","w") as f:
        json.dump(x, f)

    # Spool results to XLS file
    wb = xlwt.Workbook()
    ws = wb.add_sheet("Differences")
    ws.write(0,0,"Property")
    j = 0
    for k in x["differences"]:
        for i in range(len(x["id"])):
            if (j==0) :
                ws.write(j,i+1,f"ScanId: {x['id'][i]}")
            if (i==0) : ws.write(j+1,0,f"{k}")
            v = x["differences"][k]
            ws.write(j+1,i+1,v[i])
        j+=1
    ws = wb.add_sheet("Manifest")
    ws.write(0,0,"Path")
    j = 0
    for k in x["manifest"]:
        for i in range(len(x["id"])):
            if (j==0) :
                ws.write(j,i+1,f"Sz: {x['id'][i]}")
                ws.write(j,i+len(x["id"])+1,f"Dt: {x['id'][i]}")
            if (i==0) : ws.write(j+1,0,f"{k}")
            v = x["manifest"][k]
            ws.write(j+1,i+1,v[i])
            ws.write(j+1,i+len(x["id"])+1,v[(len(x["id"])+i)])
        j+=1
    wb.save("Differences.xls")

# Fetch comparisons of these scans
fetchScans([6395564, 6395413, 6386441])